src/Libraries/CoreNodes/Data.cs

```diff
@@ -198,9 +201,66 @@ public static string StringifyJSON([ArbitraryDimensionArrayImport] object values
 #if _WINDOWS
                     new PNGImageConverter(),
 #endif
-                });
+                }
+            };
+
+            StringBuilder sb = new StringBuilder(256);
+            using (var writer = new StringWriter(sb, CultureInfo.InvariantCulture))
+            {
+                using (var jsonWriter = new MaxDepthJsonTextWriter(writer))
+                {
+                    JsonSerializer.Create(settings).Serialize(jsonWriter, values);
+                }
+                return writer.ToString();
+            }
         }
 
+        /// <summary>
+        /// Subclass of JsonTextWriter that limits a maximum supported object depth to prevent circular reference crashes when serializing arbitrary .NET objects types.
+        /// </summary>
+        private class MaxDepthJsonTextWriter : JsonTextWriter
+        {
+            private readonly int maxDepth = 15;
+            private int depth = 0;
+
+            public MaxDepthJsonTextWriter(TextWriter writer) : base(writer) { }
```

Comment:
maybe allow setting the depth in the constructor?

Reply:
I had it that way but simplified it since we aren't using this anywhere else currently.  Right now one place to set it.